# Project C_INFO1998

In [18]:
# Project A_info1998 (clean Data)
import pandas as pd 
import numpy as np
# import dataset 
filepath = 'D:/Cornell/FALL2017/info1998/Project1/train.csv'
house = pd.read_csv(filepath)
# Baisc Clean Data in ProjectA
med = house['LotFrontage'].describe()['50%']
house['LotFrontage'] = house['LotFrontage'].fillna(value = med)
# Alley
del house['Alley']
# MasVnrArea
med = house['MasVnrArea'].describe()['50%']
house['MasVnrArea'] = house['MasVnrArea'].fillna(value = med)
# FireplaceQu
del house['FireplaceQu']
# GarageType
house['GarageType'] = house['GarageType'].fillna(value = -200)
# GarageYrBlt
house['GarageYrBlt'] = house['GarageYrBlt'].fillna(value = med)
# PoolQC
del house['PoolQC']
# Fence
del house['Fence']
# MiscFeature
del house['MiscFeature']
# fill other all na value to -200 
house_filled = house.fillna(value=-200.0)
# check all missing data 
sum(house_filled.isnull().any())

0

## Logistics Regression

In [12]:
#exclude 5 in OverallQual column
house = house[house.OverallQual !=5]

In [15]:
#do correlation test to choose predictors
house.corr()

Id  MSSubClass  LotFrontage   LotArea  OverallQual  \
Id             1.000000    0.010945    -0.021673 -0.053618    -0.032456   
MSSubClass     0.010945    1.000000    -0.421451 -0.221569    -0.032984   
LotFrontage   -0.021673   -0.421451     1.000000  0.330028     0.263559   
LotArea       -0.053618   -0.221569     0.330028  1.000000     0.123945   
OverallQual   -0.032456   -0.032984     0.263559  0.123945     1.000000   
OverallCond    0.013649   -0.021382    -0.030216  0.026116    -0.016983   
YearBuilt     -0.032828    0.014905     0.109997 -0.009272     0.546735   
YearRemodAdd  -0.031832    0.037575     0.099001  0.023058     0.576858   
MasVnrArea    -0.078346   -0.022000     0.190437  0.116356     0.383196   
BsmtFinSF1    -0.004514   -0.088034     0.232362  0.223045     0.284845   
BsmtFinSF2    -0.021243   -0.022570     0.029036  0.134000    -0.009285   
BsmtUnfSF     -0.022122   -0.211200     0.126945 -0.000329     0.278186   
TotalBsmtSF   -0.033442   -0.308345     0.378209  0.276328     0.570433   
1stFlrSF       0.004427   -0.343174     0.428914  0.304691     0.499617   
2ndFlrSF      -0.016529    0.258771     0.085621  0.061478     0.232187   
LowQualFinSF  -0.046516    0.066554     0.037265  0.006120    -0.027481   
GrLivArea     -0.014650   -0.039037     0.407614  0.288373     0.581573   
BsmtFullBath  -0.008738   -0.009596     0.100422  0.152117     0.159461   
BsmtHalfBath  -0.014317    0.030387    -0.008689  0.054591    -0.007516   
FullBath      -0.032359    0.012732     0.201847  0.125075     0.511078   
HalfBath      -0.006120    0.181084     0.042306  0.008504     0.240554   
BedroomAbvGr   0.012276   -0.133493     0.275416  0.141568     0.149472   
KitchenAbvGr  -0.028313    0.233858    -0.029117 -0.017416    -0.202410   
TotRmsAbvGrd  -0.000023   -0.082319     0.356782  0.214084     0.452225   
Fireplaces    -0.008394   -0.119678     0.256743  0.280465     0.381299   
GarageYrBlt    0.054583   -0.077428     0.110347  0.077918     0.361707   
GarageCars     0.022529   -0.136070     0.295509  0.158206     0.613460   
GarageArea     0.025863   -0.183643     0.364678  0.194180     0.590570   
WoodDeckSF    -0.032737   -0.052130     0.086598  0.138681     0.270206   
OpenPorchSF   -0.025331   -0.031870     0.135232  0.085018     0.287265   
EnclosedPorch -0.012793   -0.001662     0.033366  0.001810    -0.103526   
3SsnPorch     -0.025299   -0.054990     0.066218  0.022329     0.023902   
ScreenPorch   -0.001001   -0.024235     0.030285  0.027260     0.052144   
PoolArea       0.067167    0.003971     0.194286  0.087592     0.051266   
MiscVal       -0.021312   -0.020376     0.000472  0.016679    -0.032504   
MoSold        -0.000852   -0.006065     0.004669 -0.003131     0.104746   
YrSold         0.023830   -0.015240     0.018541 -0.015974    -0.048598   
SalePrice     -0.033775   -0.167445     0.352656  0.281635     0.768315   

               OverallCond  YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  \
Id                0.013649  -0.032828     -0.031832   -0.078346   -0.004514   
MSSubClass       -0.021382   0.014905      0.037575   -0.022000   -0.088034   
LotFrontage      -0.030216   0.109997      0.099001    0.190437    0.232362   
LotArea           0.026116  -0.009272      0.023058    0.116356    0.223045   
OverallQual      -0.016983   0.546735      0.576858    0.383196    0.284845   
OverallCond       1.000000  -0.399173      0.026158   -0.093953   -0.066226   
YearBuilt        -0.399173   1.000000      0.662227    0.284247    0.243546   
YearRemodAdd      0.026158   0.662227      1.000000    0.176583    0.150520   
MasVnrArea       -0.093953   0.284247      0.176583    1.000000    0.279086   
BsmtFinSF1       -0.066226   0.243546      0.150520    0.279086    1.000000   
BsmtFinSF2        0.064405  -0.041628     -0.072844   -0.050200   -0.016965   
BsmtUnfSF        -0.096585   0.149788      0.199298    0.086156   -0.506069   
TotalBsmtSF      -0.144161   0.389782      0.331306    0.361296    0

In [20]:
# Logistics regression
# set predictors
X = pd.DataFrame(house[['YearBuilt','YearRemodAdd','GarageCars','SalePrice']])
#tranform predictors to categorical integer values
enc = LabelEncoder()
for i in X.columns:
    X[i] = enc.fit_transform(X[i])
# set response variable Y=OverallQual  >= average as 1 and < average as 0 
Y  = (house['OverallQual'].values >= np.mean(house['OverallQual'])) + 0 

In [21]:
# import necessary packages
import pandas as pd
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression

In [29]:
#baseline calculation
#calculated as the percentage of the column's most frequent value
sum(Y == 1) > sum(Y == 0)
baseline = float(sum(Y == 1))/len(Y)
print'baseline: ', 100.0*baseline, '%'

baseline:  51.5522107244 %


In [30]:
random_state=0
for i in range(5):
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3, random_state = random_state)
    
    logit = LogisticRegression()
    logit.fit(X_train, Y_train)
    
    # make predictions on test set
    pred_logit = logit.predict(X_test)
    accuracy=accuracy_score(y_true = Y_test, y_pred = pred_logit)
    print 'Split',i+1,'accuracy rate is:',accuracy,'. Compared to the baseline, it is',(accuracy-baseline)/baseline*100,'% more accurate.'
    
    
    random_state+=1

Split 1 accuracy rate is: 0.865203761755 . Compared to the baseline, it is 67.8305837128 % more accurate.
Split 2 accuracy rate is: 0.815047021944 . Compared to the baseline, it is 58.101274512 % more accurate.
Split 3 accuracy rate is: 0.818181818182 . Compared to the baseline, it is 58.7093563371 % more accurate.
Split 4 accuracy rate is: 0.833855799373 . Compared to the baseline, it is 61.7497654623 % more accurate.
Split 5 accuracy rate is: 0.833855799373 . Compared to the baseline, it is 61.7497654623 % more accurate.


Paragraph： At first I exclude the OverallQual value that's equals to 5 as it required in the project description. Then I did correlation check to pick my predictors as the variables most correlated to my response variable, including 'YearBuilt','YearRemodAdd','GarageCars' and 'SalePrice'. In the next step, I transformed them into categorical integer values for conducting logistic regression. After that, I calculated the baseline as the percentage of the columns' most frequent value, which is 51.55%. Next, I write a for loop to conduct logistic regression five different split of train/test dataset regarding to different random states. My result turns out that logistic regression on different data split are all more than 40% accurate than baseline.

## SVM Model

In [6]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd 
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn import svm

In [7]:
## labelize 
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
le.fit(house['HouseStyle'])
## unique levels for houseStyle
print('Unique classifications:\n',list(le.classes_))
y = le.transform(house['HouseStyle'])

Unique classifications:
 [0, 1, 2, 3, 4, 5, 6, 7]


In [8]:
# baseline (it is equivalent to the percentage of the column's most frequent value.)
## count the frequency for each classifications 
print('Frequency table:\n',pd.DataFrame({'y':list(y)}).y.value_counts())
print()
baseline = round(sum(y == 2)/len(y),3)
print('Baseline:',baseline)

Frequency table:
 2    726
5    445
0    154
7     65
6     37
1     14
4     11
3      8
Name: y, dtype: int64

Baseline: 0.497


In [9]:
# MSSubClass, 2ndFlrSF ,HalfBath
dataset = house
dataset['HouseStyle'] = y
dataset.corr()['HouseStyle'] > 0.35
## choose the variables with correlation greater than 0.35. 
X = dataset[['MSSubClass','2ndFlrSF','HalfBath']]

In [10]:
for i in range(1,6):
    print('Iter',i,':\n')
    print('test_size split:', str(round(0.1 * i,2)))
    print('random_state:', i)
    classifier = svm.SVC(kernel = 'rbf',C = 1)
    ## splot the training and testing data 
    X_train, X_test, Y_train, Y_test = train_test_split(X,y,test_size = 0.1, random_state = i)
    ## fit the model to the training data 
    classifier.fit(X_train, Y_train)
    # predict the y_test 
    pred = classifier.predict(X_test)
    # accuray score
    score = classifier.score(X_test,Y_test)
    improve = (score - baseline)/baseline
    print('The accuracy score: ',score, '\n')
    print('The improvement based on Baseline: ',improve, '\n')
    print('\n\n')

Iter 1 :

test_size split: 0.1
random_state: 1
The accuracy score:  0.883561643836 

The improvement based on Baseline:  0.777790027838 




Iter 2 :

test_size split: 0.2
random_state: 2
The accuracy score:  0.917808219178 

The improvement based on Baseline:  0.846696618065 




Iter 3 :

test_size split: 0.3
random_state: 3
The accuracy score:  0.869863013699 

The improvement based on Baseline:  0.750227391748 




Iter 4 :

test_size split: 0.4
random_state: 4
The accuracy score:  0.91095890411 

The improvement based on Baseline:  0.832915300019 




Iter 5 :

test_size split: 0.5
random_state: 5
The accuracy score:  0.917808219178 

The improvement based on Baseline:  0.846696618065 






The SVM model is applied to predict the variable of HouseStyle with the several parameters. Before Building the SVM model, firstly I transformed the predicted valuable of HouseStyle to 8 catergorical numerical levels 0 to 7. The variables of MSSubClass, 2ndFlrSF ,HalfBath are chosen due the correlation with the predicted variable which the correlations are greater than 0.35. And also, I calculated the baseline of HouseStyle as 0.49. All five SVM models I built are 75% more than accuracy of Baseline.  

## Decision Tree

In [14]:
# baseline (it is equivalent to the percentage of the column's most frequent value.)
# count the frequency for each classifications 
print('Frequency table:\n',pd.DataFrame({'Y':list(Y)}).Y.value_counts())
print()
baseline = float(sum(Y == 'CollgCr'))/len(Y)
print('Baseline: %f' % baseline)

Frequency table:
 NAmes      225
CollgCr    150
OldTown    113
Edwards    100
Somerst     86
Gilbert     79
NridgHt     77
Sawyer      74
NWAmes      73
SawyerW     59
BrkSide     58
Crawfor     51
Mitchel     49
NoRidge     41
Timber      38
IDOTRR      37
ClearCr     28
SWISU       25
StoneBr     25
Blmngtn     17
MeadowV     17
BrDale      16
Veenker     11
NPkVill      9
Blueste      2
Name: Y, dtype: int64

Baseline: 0.102740


In [17]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

X = np.array(house[['OverallQual', 'YearBuilt', 'LotArea']])
Y = house['Neighborhood']

random_state = 0
for i in range(5):

    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state = random_state)

    # build model and fit on train set
    dTree = DecisionTreeClassifier(max_leaf_nodes=15)
    dTree.fit(X_train, Y_train)

    random_state += 1
    
    dTree_pred = dTree.predict(X_test)
    dTree_pred
    
    accuracy = accuracy_score(y_true = Y_test, y_pred = dTree_pred)
    print('Split',i+1,'accuracy rate is:',accuracy,'. Compared to the baseline, it is',((accuracy-baseline)/baseline) * 100,'% more accurate.')

Split 1 accuracy rate is: 0.383561643836 . Compared to the baseline, it is 273.333333333 % more accurate.
Split 2 accuracy rate is: 0.372146118721 . Compared to the baseline, it is 262.222222222 % more accurate.
Split 3 accuracy rate is: 0.376712328767 . Compared to the baseline, it is 266.666666667 % more accurate.
Split 4 accuracy rate is: 0.388127853881 . Compared to the baseline, it is 277.777777778 % more accurate.
Split 5 accuracy rate is: 0.401826484018 . Compared to the baseline, it is 291.111111111 % more accurate.


To build out the decision tree model, we have selected the following independent variables: OverallQual, YearBuilt, LotArea; all of these are critical to predict Neighborhood as an outcome. Our selection rationale is based on the relevancy of the predictor variables. For OverallQual, the level of rating helps to determine the type of neighborhood; YearBuilt is useful to categorize the time periods in which the neighborhoods were built; Lastly, the variable of LotArea might help to signal the type of neighborhood as an attribute. 
